src/Workspaces/Core/MSBuild/MSBuild/Build/ProjectBuildManager.cs

```diff
@@ -0,0 +1,296 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+using Microsoft.CodeAnalysis.MSBuild.Logging;
+using Roslyn.Utilities;
+using MSB = Microsoft.Build;
+
+namespace Microsoft.CodeAnalysis.MSBuild.Build
+{
+    internal class ProjectBuildManager
+    {
+        private static readonly XmlReaderSettings s_xmlReaderSettings = new XmlReaderSettings()
+        {
+            DtdProcessing = DtdProcessing.Prohibit,
+            XmlResolver = null
+        };
+
+        private readonly MSB.Evaluation.ProjectCollection _projectCollection;
+        private readonly MSBuildDiagnosticLogger _logger;
+        private bool _started;
+
+        public ProjectBuildManager()
+        {
+            var properties = new Dictionary<string, string>()
+            {
+                { "DesignTimeBuild", "true" }, // this will tell msbuild to not build the dependent projects
+                { "BuildingInsideVisualStudio", "true" }, // this will force CoreCompile task to execute even if all inputs and outputs are up to date
+                { "BuildProjectReferences", "false" },
+                { "BuildingProject", "false" },
+                { "ProvideCommandLineArgs", "true" }, // retrieve the command-line arguments to the compiler
+                { "SkipCompilerExecution", "true" }, // don't actually run the compiler
+                { "ContinueOnError", "ErrorAndContinue" }
+            };
+
+            _projectCollection = new MSB.Evaluation.ProjectCollection(properties);
+
+            _logger = new MSBuildDiagnosticLogger()
+            {
+                Verbosity = MSB.Framework.LoggerVerbosity.Normal
+            };
+        }
+
+        private MSB.Evaluation.Project FindProject(string path, IDictionary<string, string> globalProperties)
+        {
+            var loadedProjects = _projectCollection.GetLoadedProjects(path);
+            if (loadedProjects == null || loadedProjects.Count == 0)
+            {
+                return null;
+            }
+
+            globalProperties = globalProperties ?? ImmutableDictionary<string, string>.Empty;
+            var totalGlobalProperties = _projectCollection.GlobalProperties.Count + globalProperties.Count;
+
+            foreach (var loadedProject in loadedProjects)
+            {
+                if (loadedProject.GlobalProperties.Count != totalGlobalProperties)
```

Comment:
so is this like best effort or there should be only 1 that exactly match? or as long as property match, we don't care what project it actually is?

Reply:
We already know what project it is from the call to `_projectCollection.GetLoadedProjects(path);` above. That MSBuild API can return multiple projects for the same path if they differ by global properties. Here, we're sifting through those projects for the project which matches the global properties we expect.

Reply:
I added additional comments here to describe what's happening.